Introduce series creator field

geichelberger opened the pull request:
The database was missing a field to store the creator of a series, and it was only available in the index. After rebuilding an index, the information will be lost.

This is just a draft. If the solution is feasible, I will add a migration script.

Fixes https://github.com/opencast/opencast/issues/6428

<!--
Explain what this pull request does.  If this fixes a bug, also explain how to reproduce the issue.
-->

<!--
If your pull request is targeting the legacy (second oldest release branch), explain why this patch is so important that
it needs to go into the legacy and can not go just into stable or develop.
-->

### How to test this patch

<!--
Explain how to test this patch. If this requires a particular set-up or configuration, explain that as well and provide
the necessary configuration if possible. The easier it is for others to test the patch, the faster this will get
reviewed and merged.
-->


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

geichelberger committed:
Introduce series creator field

The database was missing a field to store the creator of a series, and
it was only available in the index. After rebuilding an index, the
information will be lost.

jduehring commented:
Tested your patch with a fresh OC 17.x installation and it worked without issues and the creator field stayed intact. 

Would like to see this being added. 

We could probably simplify a tiny bit by chaining the Series construcor right? 

```
  public Series(String identifier, String organization) {
    this(identifier, organization, null)
  }
```

jduehring commented on modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/objects/series/Series.java:
```diff
@@ -168,6 +168,20 @@ public Series(String identifier, String organization) {
+  /**
+   * The series identifier.
+   *
+   * @param identifier
+   *          the object identifier
+   * @param organization
+   *          the organization
+   */
```
```suggestion
   * @param identifier
   *          the object identifier
   * @param organization
   *          the organization
   * @param creator
   *          the creator
   */
```

geichelberger committed:
Introduce series creator value object

Added a new value object to represent the username/name tuple for the
series creator. Updated the `Series` and `SeriesEntity` class to use the
new value object for the creator field.

geichelberger committed:
Add migration scripts and guide

Signed-off-by: Gregor Eichelberger <gregor.eichelberger@tuwien.ac.at>

mtneug commented on docs/upgrade/17_to_18/extract_es_data.sh:
```diff
@@ -0,0 +1,36 @@
+
+    RESPONSE=$(curl -s -u "$ES_USER:$ES_PASS" -X POST "$ES_HOST/$INDEX/_search" -H "Content-Type: application/json" -d "$QUERY")
+
+    HITS_COUNT=$(echo "$RESPONSE" | jq '.hits.hits | length')
```
Can we assume `jq` is installed on the host? Maybe mention the CLI tools required for this script in the release notes?

mtneug commented on docs/upgrade/17_to_18/mariadb-postgresql.sql:
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE oc_series ADD creator_username VARCHAR(128);
+ALTER TABLE oc_series ADD creator_name VARCHAR(255);
+UPDATE oc_series SET creator_username = '', creator_name = '';
+ALTER TABLE oc_series MODIFY creator_username VARCHAR(128) NOT NULL;
```
```suggestion
ALTER TABLE oc_series ALTER creator_username SET NOT NULL;
```

geichelberger replied:
I used an online syntax checker, and it did not complain for PostgreSQL :see_no_evil:

mtneug commented on docs/upgrade/17_to_18/extract_es_data.sh:
```diff
@@ -0,0 +1,44 @@
+        break
+    fi
+
+    echo "$RESPONSE" | jq -r '.hits.hits[] | "UPDATE oc_series SET creator_name = \"" + (._source.creator[0] // "") + "\" WHERE id = \"" + (._source.uid[0] // "") + "\" AND organization = \"" + (._source.organization[0] // "") + "\";"' >> "$OUTPUT_FILE" 2>> "$ERROR_LOG"
```
Double quotes are not considered strings but identifiers in PostgreSQL.

geichelberger committed:
Use single quotes for UPDATE statement

Doubles quotes are used for identifiers in PostgreSQL.

geichelberger committed:
Split DB migration script

MariaDB and PostgresSQL use a different syntax for altering tables.

KatrinIhler commented:
@jduehring Could you review this again?

Also this is currently working for the University of Vienna in production.

@mtneug Were your concerns addressed?

mtneug commented:
It appears to be resolved at first glance.

jduehring commented:
Tested again, including the migration script and everything looks good and seems to work.
